Move Snapshot APIs in-tree

xing-yang opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR moves snapshot APIs in-tree and also adds a snapshot controller in-tree.
The design proposal is here: kubernetes/community#1662

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
https://github.com/kubernetes/features/issues/543

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Volume snapshot APIs are moved in-tree.  Snapshot feature is added for in-tree and out-of-tree CSI plugins. 
```

xing-yang committed:
Snapshot APIs: Add feature gate

This PR moves snapshot APIs in-tree.

xing-yang committed:
gofmt changes (Snapshot APIs)

This PR moves snapshot APIs in-tree.

xing-yang committed:
Snapshot APIs: Add generated files

This PR moves snapshot APIs in-tree.

xing-yang committed:
Snapshot APIs: add printers

This PR moves snapshot APIs in-tree.

xing-yang committed:
Move Snapshot APIs in-tree

This PR moves snapshot APIs in-tree.

xing-yang committed:
Move Snapshot APIs in-tree

This PR moves snapshot APIs in-tree and also adds
a snapshot controller in-tree.
The design proposal is here: kubernetes/community#1662

dims commented:
@kubernetes/sig-storage-api-reviews 
@kubernetes/api-reviewers 

/sig storage

dixudx commented:
I'd prefer to keep this PR and close #64365. Not need to split a single feature into two separate PRs, which will be inconvenient for integration tests.

wackxu commented:
cc

Commit:
Add generated files

Commit:
Fix test errors

Commit:
Snapshot APIs: Change VSD to VolumeSnapshotData in controller

This commit spells out VS/VSD to VolumeSnapshot/VolumeSnapshotData.

Commit:
Snapshot APIs: Add StorageClass to VolumeSnapShotSpec

This commit adds StorageClass to VolumeSnapshotSpec,
add SnapshotParameters to StorageClass, and add
VolumeSnapshotConditionCreating, etc.

jingxu97 commented on pkg/apis/storage/types.go:
```diff
@@ -80,6 +80,12 @@ type StorageClass struct {
+	// This is an optionally-prefixed name, like a label key.
+	// For example: "kubernetes.io/gce-pd" or "kubernetes.io/aws-ebs".
+	// This value may not be empty.
+	Snapshotter string
```
We may not need this if provisioner is sufficient for this purpose.

jingxu97 commented on pkg/controller/volume/snapshot/snapshot_controller.go:
```diff
@@ -0,0 +1,1470 @@
+	vol "k8s.io/kubernetes/pkg/volume"
+)
+
+// Number of retries when we create a VSD object for a created snapshot.
```
VSD --> volume snapshot data or VolumeSnapshotData.

xing-yang replied:
Sure.

rootfs commented:
/ok-to-test

rootfs commented:
@xing-yang need a release note

rootfs commented:
/assign @jsafrane @tsmetana

jingxu97 commented on pkg/controller/volume/snapshot/snapshot_controller.go:
```diff
@@ -0,0 +1,1470 @@
+		snapshotDataQueue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "csi-snapshotter-vsd"),
+	}
+
+	// Prober is nil because vsd is not aware of Flexvolume.
```
vsd --> volume snapshot

xing-yang replied:
Sure.

jingxu97 commented on pkg/controller/volume/snapshot/snapshot_controller.go:
```diff
@@ -0,0 +1,1470 @@
+	return ctrl, nil
+}
+
+// enqueueWork adds vs or vsd to given work queue.
```
vs -- >volume snapshot

xing-yang replied:
Sure.

jingxu97 commented on pkg/controller/volume/snapshot/snapshot_controller.go:
```diff
@@ -0,0 +1,1470 @@
+}
+
+type SnapshotController struct {
+	vsLister           storagelisters.VolumeSnapshotLister
```
maybe snapshotLister and snapshotDataLister?

xing-yang replied:
VolumeSnapshotLister is an auto-generated name.  The only way to change it is to rename the API object from VolumeSnapshot to Snapshot and then re-generate the Lister.

xing-yang replied:
@jingxu97 You meant to rename "vsLister" to "snapshotLister"?  Sure.

wackxu commented:
/test pull-kubernetes-integration

xing-yang commented:
The PR is moved to https://github.com/kubernetes-csi/external-snapshotter/pull/7.  Close this one.
